### ISSUE TITLE
no-unbound-method: Disable for decorated methods

### ISSUE THREAD

**Issue by csvn** [palantir/tslint#3950]
## Feature request

Allow decorator to disable `no-unbound-this` rule for a specific method.

```typescript
// greeter.ts
import { bind } from 'decorators';

class Greeter {
  name = 'John';

  @bind
  greet() { console.log(`Hello ${this.name}!`); }
}

const greeter = new Greeter();
const greet = greeter.greet;
greet();
// Should not error, `@bind` performs `.bind()` for us!

// tslint.json
{
  "rules": {
    "no-unbound-this": [true, ["bind", "autobind"]]
  }
}
```

It would be great if it was possible to adjust `no-unbound-method` so that it won't error if a method is decorated with a function with a specified name. There are existing libraries (e.g. [core-decorators/autobind](https://github.com/jayphelps/core-decorators#autobind)) that has functionality like this. Unfortunately it makes `no-unbound-method` almost unusable.

Would it be possible to do a PR for this?

**Comment by giladgray** [comment]
this seems like an edge case that actually highlights a poor coding practice: use the fat arrow syntax, which automatically binds and is gloriously supported in the language itself.

`greet = () => console.log(`Hello ${this.name}!`)`

**Comment by csvn** [comment]
Makes sense, that's a decent workaround. I can buy not having this feature in tslint core (might be better as a custom rule if required).

I don't agree fully about poor coding practice though. As always, Javascript style is highly subjective.

- A decorator can bind lazily, all property arrow functions are instantiated eagerly (i.e. in the constructor)
- The methods actually look like methods instead of properties
- With "member-ordering" the bound methods end up next to other methods

Anyhow, thanks for the response.

**Comment by karthikv** [comment]
@csvn I agree, I think there are times where you'd prefer a decorator (e.g. if you need to call the method with `super`, if you need the function to have a non-empty `name` property, and for the reasons you wrote). I wrote a custom rule that doesn't complain if the unbound method has a decorator that you specify: https://github.com/karthikv/binding_rules (see `no-unbound-method-decorator`). I also added a rule `no-arrow-class-property` to disallow arrow functions as class properties, thereby letting you enforce the use of a decorator for all bound methods.
